### PR TITLE
feat: added SDK macros edge cases test coverage

### PR DESCRIPTION
**Pull request by chefsale**
# SDK macros edge cases test coverage

## Description

This PR addresses the bounty "Add tests for SDK macros edge cases" by significantly expanding the test suite for `crates/sdk/macros/src/`. The goal is to improve the robustness and error reporting of the SDK macros by covering various complex and edge-case scenarios.

New tests have been added to `crates/sdk/tests/macros/` to cover:
- **Nested Types:** Including nested `Option`, `Vec` of structs, deeply nested types, tuples, and arrays.
- **Complex Generics:** Valid usage of multiple lifetime parameters, static lifetimes, lifetime elision, slice references, and `Option`/`Result` with lifetimes.
- **Error Message Quality:** Specific compile-fail tests for unsupported patterns such as incorrect `init` method naming/attributes, private events, trait implementations, const generics, explicit ABIs, and `impl Trait` arguments.
- **Interaction with other attributes:** Verifying compatibility of `#[app::*]` macros with standard Rust attributes like `#[derive]`, `#[doc]`, `#[allow]`, `#[repr]`, `#[cfg]`, `#[must_use]`, `#[inline]`, and `#[deprecated]`.
- **Valid Events:** Comprehensive tests for various event definitions including simple, data, lifetime, nested, tuple, and mixed variants.

## Test plan

The new test files are integrated into the existing macro test suite.
- To verify the changes, navigate to the `crates/sdk` directory and run `cargo test`.
- The compile-fail tests will ensure that unsupported patterns correctly produce the expected error messages.
- `cargo fmt --check` was run to ensure all new code adheres to the project's formatting standards.
- No user-interface changes were made, so no screenshots are applicable.
- These are macro-level tests and do not directly map to end-to-end test cases.

## Documentation update

No public or internal documentation updates are required as these changes are internal test improvements.

---
<a href="https://cursor.com/background-agent?bcId=bc-aba96a0a-4e00-4734-b3dc-90545a069e96"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/open-in-cursor-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/open-in-cursor-light.png">Open in Cursor</picture></a>&nbsp;<a href="https://cursor.com/agents?id=bc-aba96a0a-4e00-4734-b3dc-90545a069e96"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/open-in-web-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/open-in-web-light.png">Open in Web</picture></a>

---

> [!NOTE]
> **Low Risk**
> Low risk because changes are limited to compile-time test fixtures and expected stderr outputs, with no runtime or production logic modifications.
> 
> **Overview**
> **Significantly expands SDK macro test coverage** by adding new `trybuild` pass cases for nested/complex types, advanced lifetime patterns, attribute interactions (e.g. `#[derive]`, `#[cfg]`, `#[repr]`, `#[deprecated]`), and valid `#[app::event]` enum shapes.
> 
> **Adds focused compile-fail fixtures** to lock in error-message quality for unsupported patterns (mis-specified `#[app::init]`, private events, trait impls, `impl Trait` args, const generics, explicit ABIs, and `#[app::event]` on a struct), and updates existing `.stderr` expectations to match current diagnostics/wording.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 3ab127cc26e35798f4c84b55c63a185cff4d95a3. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: crates/sdk/tests/macros/attribute_interactions.rs [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+    pub fn get_y(&self) -> i32 {
+        self.y
+    }
+}
+
+// Test cfg attributes work with state (basic)
+#[app::state]
+struct ConditionalState {
+    #[cfg(feature = "test_feature")]
```
Comment (xilosada) [inline crates/sdk/tests/macros/attribute_interactions.rs]: 🟡 **Conditional field uses non-existent feature flag**

The test uses `#[cfg(feature = "test_feature")]` but this feature is not defined in the crate's Cargo.toml. The field `conditional_field` will never be compiled, so this test doesn't actually verify that `#[cfg]` attributes work correctly with the SDK macros - it only tests the case where the conditional code is excluded.

**Suggested fix:**
```
Either add `test_feature` to the crate's Cargo.toml features and run tests with `--features test_feature`, or add a separate test case using a cfg predicate that evaluates to true (e.g., `#[cfg(test)]`) to verify both inclusion and exclusion scenarios.
```

File: crates/sdk/tests/macros/valid_events.rs [hunk 2]
```diff
@@ -0,0 +1,61 @@
+//! Tests for valid event definitions with various type patterns
```
Comment (xilosada) [inline crates/sdk/tests/macros/valid_events.rs:1]: 💡 **Missing compile-fail tests for invalid event patterns**

The PR adds comprehensive valid event tests but only one compile-fail test for events (`error_private_event.rs`). Other invalid event patterns like events with generic type parameters (not lifetimes), events applied to structs instead of enums, or events with unsupported field types are not covered.

**Suggested fix:**
```
Consider adding compile-fail tests for: `#[app::event]` on a struct, events with type generic parameters like `enum Event<T>`, and events with field types that don't implement required traits.
```

File: crates/sdk/tests/macros/nested_types.rs [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        &self.option_vec
+    }
+
+    pub fn get_vec_option(&self) -> &Vec<Option<String>> {
+        &self.vec_option
+    }
+}
+
+// Test nested struct types
+struct InnerData {
```
Comment (xilosada) [inline crates/sdk/tests/macros/nested_types.rs:39]: 💡 **Non-state structs lack serialization context**

The `InnerData` and `MiddleLayer` structs are used within state fields but have no derives or documentation explaining serialization requirements. While this tests macro expansion, users copying this pattern may miss that these types need to implement serialization traits (like `BorshSerialize`/`BorshDeserialize`) for actual runtime use.

**Suggested fix:**
```
Add a comment clarifying that in production use, these nested types would need appropriate serialization derives, or add the derives to make the example more complete.
```

File: crates/sdk/tests/macros/error_private_event.stderr [hunk 4]
```diff
@@ -0,0 +1,5 @@
+error: (calimero)> this event must be public
+ --> tests/macros/error_private_event.rs:6:6
```
Comment (xilosada) [inline crates/sdk/tests/macros/error_private_event.stderr:2]: 📝 **Inconsistent stderr formatting may cause test brittleness**

The arrow formatting uses single space (` -->`) while other stderr files like error_init_not_named.stderr use double space (`  -->`). Trybuild performs exact string matching on stderr output, so if the actual compiler output differs from what's recorded here, the test will fail. This inconsistency suggests the stderr files may have been hand-written rather than captured from actual compiler output.

**Suggested fix:**
```
Run the actual tests to capture the real compiler stderr output, or verify the spacing matches what the macros actually produce.
```

File: crates/sdk/tests/macros/invalid_nested_generics.stderr [hunk 5]
```diff
@@ -0,0 +1,5 @@
+error: (calimero)> generic types are not supported
+ --> tests/macros/invalid_nested_generics.rs:9:6
```
Comment (xilosada) [inline crates/sdk/tests/macros/invalid_nested_generics.stderr:2]: 📝 **Inconsistent stderr formatting may cause test brittleness**

Same issue as error_private_event.stderr - uses single space before arrow (` -->`) while most other stderr files use double space (`  -->`). This inconsistency in expected error output format could cause test failures if the compiler output doesn't exactly match.

**Suggested fix:**
```
Verify the expected stderr matches actual compiler output by running the trybuild tests.
```

File: crates/sdk/tests/macros.rs [hunk 6]
```diff
@@ -11,11 +11,43 @@ fn all() {
 
     // todo! break these up into pass/fail dirs for organization
```
Comment (xilosada) [inline crates/sdk/tests/macros.rs:13]: 📝 **TODO comment about directory organization not addressed**

The existing TODO comment suggests breaking tests into pass/fail directories for organization, but this PR adds 14 new test files in the flat structure. While the added section comments improve readability, following through on the TODO would provide better long-term maintainability.

**Suggested fix:**
```
Consider organizing into `tests/macros/pass/` and `tests/macros/fail/` directories as suggested by the TODO, or update the TODO comment to reflect the current organizational strategy with section comments.
```

File: crates/sdk/tests/macros/error_private_event.stderr [hunk 7]
```diff
@@ -0,0 +1,5 @@
+error: (calimero)> this event must be public
```
Comment (xilosada) [inline crates/sdk/tests/macros/error_private_event.stderr:1]: 📝 **Inconsistent stderr file formatting**

This stderr file has slightly different spacing in the error output (single space before `-->` on line 2, whereas other stderr files have no leading space). While trybuild may normalize this, inconsistent expected outputs can cause brittle tests across compiler versions.

**Suggested fix:**
```
Verify the actual compiler output and ensure all stderr files use consistent formatting. Consider regenerating stderr files using `TRYBUILD=overwrite cargo test` to ensure consistency.
```
